src/main/java/org/dependencytrack/util/NotificationUtil.java

```diff
@@ -108,6 +110,29 @@ public static void analyzeNotificationCriteria(QueryManager qm, Vulnerability vu
         }
     }
 
+    public static void analyzeNotificationCriteria(QueryManager qm, Vulnerability vulnerability, VulnerabilityUpdateDiff vulnerabilityUpdateDiff) {
+        // Vulnerabilities are not guaranteed to include component relationships depending on where the event was generated
+        final Vulnerability completeVulnerability = qm.getVulnerabilityByVulnId(vulnerability.getSource(), vulnerability.getVulnId());
+        final List<Component> components = completeVulnerability.getComponents();
+        if (components != null && !components.isEmpty()) {
+            // To reduce noise we only emit a single notification for each updated vulnerability if it affects one
+            // of our components. The component details are still useful for event consumers, so we pick the first one.
+            final Component detachedComponent = qm.detach(Component.class, components.get(0).getId());
```

Comment:
I feel like this could be a bit misleading. A vulnerability can potentially affect multiple libraries, operating systems, or hardware components. What is the benefit of including only one component?

Reply:
There's definitely some nuance that I missed here. 

I considered the earlier implementation where the event notification would include 1 vulnerability, 1 component and a list of affected projects (similar to a `NEW_VULNERABILITY` notification), which would mean generating a notification for every affected component entry in the database. As each logical component has a unique entry in the database for every project that uses it, this could mean emitting a large number of notifications where the content is identical except for the `component.uuid` field. So to try and reduce this redundancy and (in theory) reduce the database load, my logic was to choose just one of the "identical" component entries and emit a single event per updated vulnerability.

What I missed was that there could be separate logical components affected by the same vulnerability and picking only one of them doesn't make sense as you said.

I think what this comes down to is what useful contextual information about a vulnerability do we want to include in the notification? For my use case I actually only care about the `vulnerability` object itself, but others may have use cases where they need the affected component information (and potentially affected projects) as well.

I may be missing something, but at the moment there doesn't seem to be a way to fetch a list of unique logical component identities without doing a full table scan and de-duplicating. If this can't be done efficiently, perhaps it makes sense not to include component information in the notification. What do you think?

Reply:
IMO we should keep the information in this notification as minimal as possible. All data we add now, users will start depending on it, and we can't easily remove it again. I think it's better to wait for users to come forward with their use cases, and then evaluating if / how that can be achieved with more data.

Transmitting affected projects is not feasible (see for example https://github.com/DependencyTrack/hyades/issues/467 where we ran into this in another context).

Communicating logical components is complicated as you said, and I am hesitant to add this complexity without us knowing that users really need it.

If we only include the vulnerability itself now, receivers will have enough information to query the `/api/v1/source/{source}/vuln/{vuln}/projects` endpoint to fetch affected projects.